Source/MQTTnet/MqttApplicationMessage.cs

```diff
@@ -29,5 +29,7 @@ public class MqttApplicationMessage
         public byte[] CorrelationData { get; set; }
 
         public List<uint> SubscriptionIdentifiers { get; set; }
+
+        public uint DeliveryCount { get; set; }
```

Comment:
Please do not extend the application message because this message only contains MQTT spec relevant stuff.
You can move this counter to the EventArgs of the undelivered messages event handler. Then you only need a variable to count these things.

Reply:
Nice Idea!